crates/iota-framework/packages/iota-system/sources/validator_set.move

```diff
@@ -1129,4 +1490,153 @@ module iota_system::validator_set {
+
+            i = i + 1;
+        };
+
+        // Sort indices based on the stake values 
+        let mut i = 1;
+        while (i < validators_num) {
+            let cur = validator_stakes[i];
+            let mut j = i;
+            while (j > 0 && validator_stakes[j-1] < cur) {
+                validator_stakes.swap(j, j - 1);
+                validator_indices.swap(j, j - 1);
+                j = j - 1;      
+            };
+            i = i + 1;
+        };
+
+        // Return the top N indices
+        let mut top_n_indices = vector[];
+        let mut i = 0;
+        while (i < n.min(validators_num)) {
+            top_n_indices.push_back(validator_indices[i]);
+            i = i + 1;
+        };
+
+        self.committee_members = top_n_indices;
+    }
+
+    // Emits events for committee validators that were added and left the committee.
+    public(package) fun emit_committee_change_events(self: &ValidatorSetV2, prev_committee_addresses: vector<address>, ctx: &TxContext) {        
+        let committee_members_num = self.committee_members.length();
+        let active_validators_num = self.active_validators.length();
+
+        let new_epoch = ctx.epoch() + 1;
+
+        
+        let mut i = 0;
+        while (i < committee_members_num) {
+            let validator_index = self.committee_members[i];
+
+            i = i + 1;
+
+            assert!(
+                validator_index < active_validators_num,
+                ECommitteeMembersSetCorrupt,
+            );
+            let validator = &self.active_validators[validator_index];
+            let validator_address = validator.iota_address();
+
+
+            // Emit join committee event only if the validator wasn't part of the old committee.
+            if (!prev_committee_addresses.contains(&validator_address)) {
+                event::emit(
+                    CommitteeValidatorJoinEvent {
+                        epoch: new_epoch,
+                        validator_address: validator_address,
+                        staking_pool_id: staking_pool_id(validator),
+                    }
+                );
+            };
+        };
+
+        // Emit leave committee events.
+        let prev_committee_num = prev_committee_addresses.length();
+        let new_committee_addresses = self.committee_validator_addresses();
+        let mut i = 0;
+        while (i < prev_committee_num) {
+            let validator_address = prev_committee_addresses[i];
+
+            i = i + 1;
+
+            // Emit leave committee event only if validator is not part of the new committee AND is still an active validator.
+            if (!new_committee_addresses.contains(&validator_address)) {
+                let mut validator_index_opt = find_validator(&self.active_validators, validator_address);
+                
+                // If it's not part of active validators anymore, it means that the leave committee event has been emitted before.
```

Comment:
Is that really the case? Can't it happen that the validator somehow gets removed from the `active_validators` set in another way? Maybe by deregistering as a validator or punishment by other validators? Idk.